### ISSUE TITLE
Option to permit no-useless-return inside blocks

### ISSUE THREAD

**Issue by felixsanz** [eslint/eslint#7785]
As discussed in https://github.com/airbnb/javascript/issues/1210, it will be great if `no-useless-return` offers an option to allow `return` **inside blocks**, to benefit consistency between blocks when using a useless `return` to quit the function execution at that point.

Here is an example:

```javascript
async function f(p) {
  if (p === 'a') {
    // some lines of code
    await fn()
    
    return
  }
  
  if (p === 'b') {
    // some lines of code
    await fn()
    
    return
  }

  if (p === 'c') {
    // some lines of code
    await fn()
    
    return // <-- option to allow tihs one
  }
}
```

Having them separated by blank lines make it clearer and easy to maintain instead of having a big `if-else if-else` block.

Having this last useless return **enforces consistency** with other blocks and also, if you add a new block below, you will **not have to modify** the (until now) last one block to add a `return` into it.

**Comment by not-an-aardvark** [comment]
Thanks for the proposal.

I'm not sure I see the benefit of having this option. With the option, the rule would *only* report things like

```js
function foo() {
  doThings();
  return;
}
```

In my opinion, it seems like this would neuter much of the benefit of using this rule. If you have a place in your code where you want to maintain consistency like this, you could always use an `// eslint-disable-line` comment and still give yourself the benefit of checking in blocks in the rest of the codebase, rather than limiting the effectiveness of the rule throughout the whole codebase.

Given that you're dealing with conditional logic, I think an `if/else` block would be better for what you're doing here (and if you're concerned about readability you could extract the logic in the block into separate functions).

So I'm not really in favor of this option, but I won't oppose it if it gets support from other members of the team.

**Comment by felixsanz** [comment]
> and if you're concerned about readability you could extract the logic in the block into separate functions

But don't you think this is so condensed?

```javascript
  if (p === 'a') {
    await refactoredFn1()
  } else if (p === 'b') {
    await refactoredFn2()
  } else if (p === 'c') {
    await refactoredFn3()
  } else if (p === 'd') {
    await refactoredFn4()
  } else if (p === 'e') {
    await refactoredFn5()
  }
```

If you want to search for a specific block it's harder.

Even if refactored into functions, using useless `return` in the last block is useful:

```javascript
  if (p === 'a') {
    await refactoredFn1()
    return
  }

  if (p === 'b') {
    await refactoredFn2()
    return
  }

  if (p === 'c') {
    await refactoredFn3()
    return
  }

  if (p === 'd') {
    await refactoredFn4()
    return
  }

  if (p === 'e') {
    await refactoredFn5()
    return
  }
```

The indentation and empty lines really help

**Comment by not-an-aardvark** [comment]
> But don't you think this is so condensed?

I don't really see the problem myself, but you could resolve that by adding linebreaks if necessary:

```js
  if (p === 'a') {
    await refactoredFn1()

  } else if (p === 'b') {
    await refactoredFn2()

  } else if (p === 'c') {
    await refactoredFn3()

  } else if (p === 'd') {
    await refactoredFn4()

  } else if (p === 'e') {
    await refactoredFn5()
  }
```

> Even if refactored into functions, using useless return in the last block is useful:

But if you forget to add an explicit `return` statement in any of the blocks, your code will silently do the wrong thing. Personally, I think having multiple `if` blocks with `return` statements in each block is much harder to read than just using `if/else` directly, because to look at that code, I have to scan through each individual `if` statement to check if it uses `return`, and then I have to wonder what happens if none of the conditions match (it seems like nothing happens in your example, but it would be unclear whether this behavior is intentional).

**Comment by platinumazure** [comment]
In my opinion, this is exactly the whole point of the rule- if you want the last return not to be flagged, don't use the rule. So I guess I would suggest using a disable comment in this case.

**Comment by kaicataldo** [comment]
> it will be great if no-useless-return offers an option to allow return inside blocks

> Having this last useless return **enforces consistency** with other blocks and also, if you add a new block below, you will **not have to modify** the (until now) last one block to add a `return` into it.

I'm a little confused here. It sounds like the proposal is to add an option that relaxes the check for the final `return` in blocks, which doesn't actually enforce consistency - it just allows it.

Also, when you mention blocks are you just referring to `if`/`else` blocks? Either way, this feels like it's going exactly against the purpose of this rule to me. I agree with @platinumazure and @not-an-aardvark and think `eslint-disable` comments should be used or the rule should be disabled entirely.

**Comment by felixsanz** [comment]
Well yeah, it's kinda the same. Allowing the last return actually enforce consistency across all blocks.

In my case i'm just talking about `if` but i think all blocks apply here. Anyway with your opinions maybe a disable comment is the best approach here :disappointed:

**Comment by not-an-aardvark** [comment]
It looks like there is consensus among the team that it would be better to just disable the rule or use disable comments here, so I'm going to close this issue.

**Comment by kaicataldo** [comment]
@felixsanz We appreciate the suggestion regardless - thanks for contributing!
